include/zephyr/drivers/stepper_motor.h

```diff
@@ -0,0 +1,461 @@
+	enum position_format format;
+	/** allows the user to specify position type such as Min, Max, Actual, Target */
+	enum position_type type;
+	/** position value */
+	int32_t position;
+};
+
+/**
+ * @brief Stepper Motor Run Info required while driving Motor in Freewheeling Mode.
+ *
+ * Here is an example of how to use this struct:
+ *
+ *      const struct stepper_motor_run_info target_position_info = {
+ *	            .direction = POSITIVE, .velocity = 200};
+ *      stepper_motor_run(my_motor, &target_position_info);
+ */
+struct stepper_motor_run_info {
+	enum run_direction direction;
+	int32_t velocity;
+};
+
+struct stepper_motor {
+	const struct device *const controller;
+	const uint8_t motor_index;
+};
+
+/**
+ * @typedef stepper_motor_set_position_t
+ * @brief Set the position of a given stepper motor device
+ *
+ * @see stepper_motor_set_position for more information
+ */
+typedef int32_t (*stepper_motor_set_position_t)(
+	const struct stepper_motor *motor,
+	const struct stepper_motor_position_info *position_info);
+
+/**
+ * @typedef stepper_motor_get_position_t
+ * @brief Get the position of a given stepper motor device
+ *
+ * @see stepper_motor_get_position for more information
+ */
+typedef int32_t (*stepper_motor_get_position_t)(const struct stepper_motor *motor,
+						struct stepper_motor_position_info *position_info);
+
+/**
+ * @typedef stepper_motor_reset_t
+ * @brief Reset the stepper motor device in a user-specific manner
+ *
+ * @see stepper_motor_reset for more information
+ */
+typedef int32_t (*stepper_motor_reset_t)(const struct stepper_motor *motor);
+
+/**
+ * @typedef stepper_motor_run_t
+ * @brief Run a.k.a do Freewheeling of a given stepper motor device
+ *
+ * @see stepper_motor_run for more information
+ */
+typedef int32_t (*stepper_motor_run_t)(const struct stepper_motor *motor,
+				       const struct stepper_motor_run_info *run_info);
+
+/**
+ * @typedef stepper_motor_stop_t
+ * @brief Stop a given stepper motor device
+ *
+ * @see stepper_motor_stop for more information
+ */
+typedef int32_t (*stepper_motor_stop_t)(const struct stepper_motor *motor);
+
+/**
+ * @typedef stepper_motor_get_stall_status_t
+ * @brief Get the stall status of a given stepper motor device
+ *
+ * @see stepper_motor_get_stall_status for more information
+ */
+typedef int32_t (*stepper_motor_get_stall_status_t)(const struct stepper_motor *motor,
+						    bool *stall_status);
+
+/**
+ * @typedef stepper_motor_calibrate_t
+ * @brief Calibrate a given stepper motor device
+ *
+ * @see stepper_motor_calibrate for more information
+ */
+typedef int32_t (*stepper_motor_calibrate_t)(const struct stepper_motor *motor);
+
+typedef int32_t (*stepper_motor_calibrate_func_t)(const struct stepper_motor *motor,
+						  int32_t *min_pos, int32_t *max_pos);
+/**
+ * @typedef stepper_motor_register_calibrate_func_t
+ * @brief Register a calibration function for a given stepper motor device
```

Comment:
A callback `calibrate_func` would have to be registered using this function. Once registered, upon calling `stepper_motor_calibrate`the pre-registered `calibrate_func`would be called. The approach is comparable to https://docs.zephyrproject.org/latest/hardware/peripherals/gpio.html#c.gpio_init_callback

Reply:
@fabiobaltieri : The idea behind this function is to perform referencing actions. Every time the motor controller (say, in this case, the trinamic 5041) starts up, the position is unknown. Ideally, one would travel to both the end stops, and say, on reaching the lower end stop, set the coordinate system, by resetting the current position to 0. Since, this is very hardware dependent (cause, one could either travel to both the end stops and detect a block, or using limit switches for detecting the end-stop), this referencing routine (here called as the ´calibrate´ function) is provided.

Reply:
sorry I still don't understand the whole mechanics of this, reading the driver code now, `tmc5041_motor_set_position` can set `MOTOR_POSITION_MIN` and `MOTOR_POSITION_MAX`, then `tmc5041_motor_register_calibrate_func` registers a callback but that only gets called from `tmc5041_motor_calibrate`, and all three are API functions so they are called by application? Or maybe a motion control algorithm, anyway I don't quite see the point, seems like a very complicated way of getting back the data that the application knew in the first place, or maybe it's just a bad example and there's cases where you read back the information from the controller? but then why bother with the callback, just return the data directly from the calibrate function